File: backend/main/java/de/griefed/serverpackcreator/versionmeta/fabric/FabricMeta.java
/* Copyright (C) 2022  Griefed
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301
 * USA
 *
 * The full license can be found at https:github.com/Griefed/ServerPackCreator/blob/main/LICENSE
 */
package de.griefed.serverpackcreator.versionmeta.fabric;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.google.common.collect.Lists;
import de.griefed.serverpackcreator.utilities.common.Utilities;
import de.griefed.serverpackcreator.versionmeta.Meta;
import java.io.File;
import java.io.IOException;
import java.net.URL;
import java.util.HashMap;
import java.util.List;
import java.util.Optional;
import javax.xml.parsers.ParserConfigurationException;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;
import org.xml.sax.SAXException;

/**
 * Fabric meta containing information about available Quilt versions and installers.
 *
 * @author Griefed
 */
public final class FabricMeta implements Meta {

  private final FabricLoader FABRIC_LOADER;
  private final FabricLoaderDetails FABRIC_LOADER_DETAILS;
  private final FabricInstaller FABRIC_INSTALLER;
  private final FabricIntermediaries FABRIC_INTERMEDIARIES;
  private final HashMap<String, FabricDetails> LOADER_DETAILS = new HashMap<>(100);

  /**
   * Create a new Fabric Meta instance, giving you access to available loader and installer
   * versions, as well as URLs to installer for further processing.
   *
   * @param fabricManifest               Fabric manifest file.
   * @param fabricInstallerManifest      Fabric-installer manifest file.
   * @param injectedFabricIntermediaries Fabric Intermediary instance.
   * @param objectMapper                 Object mapper for JSON parsing.
   * @param utilities                    Commonly used utilities across ServerPackCreator.
   * @author Griefed
   */
  public FabricMeta(
      @NotNull File fabricManifest,
      @NotNull File fabricInstallerManifest,
      @NotNull FabricIntermediaries injectedFabricIntermediaries,
      @NotNull ObjectMapper objectMapper,
      @NotNull Utilities utilities) {

    FABRIC_LOADER_DETAILS = new FabricLoaderDetails(objectMapper);
    FABRIC_LOADER = new FabricLoader(fabricManifest, utilities);
    FABRIC_INTERMEDIARIES = injectedFabricIntermediaries;
    FABRIC_INSTALLER = new FabricInstaller(fabricInstallerManifest, utilities);
  }

  @Override
  public void update() throws ParserConfigurationException, IOException, SAXException {
    FABRIC_LOADER.update();
    FABRIC_INSTALLER.update();
  }

  @Contract(pure = true)
  @Override
  public @NotNull String latestLoader() {
    return FABRIC_LOADER.latestLoaderVersion();
  }

  @Contract(pure = true)
  @Override
  public @NotNull String releaseLoader() {
    return FABRIC_LOADER.releaseLoaderVersion();
  }

  @Contract(pure = true)
  @Override
  public @NotNull String latestInstaller() {
    return FABRIC_INSTALLER.latestInstallerVersion();
  }

  @Contract(pure = true)
  @Override
  public @NotNull String releaseInstaller() {
    return FABRIC_INSTALLER.releaseInstallerVersion();
  }

  @Contract(pure = true)
  @Override
  public @NotNull List<String> loaderVersionsListAscending() {
    return FABRIC_LOADER.loaders();
  }

  @Override
  public @NotNull List<String> loaderVersionsListDescending() {
    return Lists.reverse(FABRIC_LOADER.loaders());
  }

  @Override
  public @NotNull String @NotNull [] loaderVersionsArrayAscending() {
    return FABRIC_LOADER.loaders().toArray(new String[0]);
  }

  @Override
  public @NotNull String @NotNull [] loaderVersionsArrayDescending() {
    return Lists.reverse(FABRIC_LOADER.loaders()).toArray(new String[0]);
  }

  @Contract(pure = true)
  @Override
  public @NotNull List<String> installerVersionsListAscending() {
    return FABRIC_INSTALLER.installers();
  }

  @Override
  public @NotNull List<String> installerVersionsListDescending() {
    return Lists.reverse(FABRIC_INSTALLER.installers());
  }

  @Override
  public @NotNull String @NotNull [] installerVersionsArrayAscending() {
    return FABRIC_INSTALLER.installers().toArray(new String[0]);
  }

  @Override
  public @NotNull String @NotNull [] installerVersionsArrayDescending() {
    return Lists.reverse(FABRIC_INSTALLER.installers()).toArray(new String[0]);
  }

  @Contract(pure = true)
  @Override
  public @NotNull URL latestInstallerUrl() {
    return FABRIC_INSTALLER.latestInstallerUrl();
  }

  @Contract(pure = true)
  @Override
  public @NotNull URL releaseInstallerUrl() {
    return FABRIC_INSTALLER.releaseInstallerUrl();
  }

  @Override
  public boolean isInstallerUrlAvailable(@NotNull String fabricVersion) {
    return Optional.ofNullable(FABRIC_INSTALLER.meta().get(fabricVersion)).isPresent();
  }

  @Override
  public @NotNull Optional<URL> getInstallerUrl(@NotNull String fabricVersion) {
    return Optional.ofNullable(FABRIC_INSTALLER.meta().get(fabricVersion));
  }

  @Override
  public boolean isVersionValid(@NotNull String fabricVersion) {
    return FABRIC_LOADER.loaders().contains(fabricVersion);
  }

  @Override
  public boolean isMinecraftSupported(@NotNull String minecraftVersion) {
    return FABRIC_INTERMEDIARIES.getIntermediary(minecraftVersion).isPresent();
  }

  /**
   * Get the {@link URL} to the Fabric launcher for the specified Minecraft and Fabric version.
   *
   * @param minecraftVersion Minecraft version.
   * @param fabricVersion    Fabric version.
   * @return URL to the Fabric launcher for the specified Minecraft and Fabric version.
   * @author Griefed
   */
  public @NotNull Optional<URL> improvedLauncherUrl(@NotNull String minecraftVersion,
                                                    @NotNull String fabricVersion) {
    return FABRIC_INSTALLER.improvedLauncherUrl(minecraftVersion, fabricVersion);
  }

  /**
   * Get details for a Fabric loader.
   *
   * @param minecraftVersion Minecraft version.
   * @param fabricVersion    Fabric version.
   * @return Details of a Fabric loader for the given Minecraft and Fabric version, wrapped in an
   * {@link Optional}.
   * @author Griefed
   */
  public @NotNull Optional<FabricDetails> getLoaderDetails(@NotNull String minecraftVersion,
                                                           @NotNull String fabricVersion) {
    String key = minecraftVersion + "-" + fabricVersion;

    if (LOADER_DETAILS.containsKey(key)) {

      return Optional.of(LOADER_DETAILS.get(key));

    } else if (FABRIC_LOADER_DETAILS.getDetails(minecraftVersion, fabricVersion).isPresent()) {

      LOADER_DETAILS.put(
          key, FABRIC_LOADER_DETAILS.getDetails(minecraftVersion, fabricVersion).get());
      return Optional.of(LOADER_DETAILS.get(key));

    } else {

      return Optional.empty();
    }
  }
}

File: serverpackcreator-api/src/jvmMain/kotlin/de/griefed/serverpackcreator/api/versionmeta/fabric/FabricMeta.kt
/* Copyright (C) 2023  Griefed
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301
 * USA
 *
 * The full license can be found at https:github.com/Griefed/ServerPackCreator/blob/main/LICENSE
 */
package de.griefed.serverpackcreator.api.versionmeta.fabric

import com.fasterxml.jackson.databind.ObjectMapper
import de.griefed.serverpackcreator.api.utilities.common.Utilities
import de.griefed.serverpackcreator.api.utilities.common.createDirectories
import de.griefed.serverpackcreator.api.versionmeta.Meta
import org.xml.sax.SAXException
import java.io.File
import java.io.IOException
import java.net.URL
import java.util.*
import javax.xml.parsers.ParserConfigurationException

/**
 * Fabric meta containing information about available Quilt versions and installers.
 *
 * @param fabricManifest          Fabric manifest file.
 * @param fabricInstallerManifest Fabric-installer manifest file.
 * @param fabricIntermediaries    Fabric Intermediary instance.
 * @param objectMapper            Object mapper for JSON parsing.
 * @param utilities               Commonly used utilities across ServerPackCreator.
 * @param installerCacheDirectory The cache-directory for all installers.
 *
 * @author Griefed
 */
@Suppress("unused")
actual class FabricMeta(
    fabricManifest: File,
    fabricInstallerManifest: File,
    private val fabricIntermediaries: FabricIntermediaries,
    objectMapper: ObjectMapper,
    private val utilities: Utilities,
    installerCacheDirectory: File
) : Meta {
    private val fabricLoader: FabricLoader = FabricLoader(fabricManifest, utilities)
    private val fabricLoaderDetails: FabricLoaderDetails = FabricLoaderDetails(objectMapper)
    private val fabricInstaller: FabricInstaller = FabricInstaller(fabricInstallerManifest, utilities)
    private val installerDirectory: File = File(installerCacheDirectory, "fabric")
    private val launchersDirectory: File = File(installerDirectory, "launchers")

    init {
        installerDirectory.createDirectories(create = true, directory = true)
        launchersDirectory.createDirectories(create = true, directory = true)
    }

    @Suppress("MemberVisibilityCanBePrivate")
    val loaderDetails = HashMap<String, FabricDetails>(100)

    /**
     * @author Griefed
     */
    @Throws(ParserConfigurationException::class, IOException::class, SAXException::class)
    override fun update() {
        fabricLoader.update()
        fabricInstaller.update()
    }

    /**
     * @author Griefed
     */
    actual override fun latestLoader() = fabricLoader.latestLoaderVersion()

    /**
     * @author Griefed
     */
    actual override fun releaseLoader() = fabricLoader.releaseLoaderVersion()

    /**
     * @author Griefed
     */
    actual override fun latestInstaller() = fabricInstaller.latestInstallerVersion()

    /**
     * @author Griefed
     */
    actual override fun releaseInstaller() = fabricInstaller.releaseInstallerVersion()

    /**
     * @author Griefed
     */
    actual override fun loaderVersionsListAscending() = fabricLoader.loaders

    /**
     * @author Griefed
     */
    actual override fun loaderVersionsListDescending() = fabricLoader.loaders.reversed()

    /**
     * @author Griefed
     */
    actual override fun loaderVersionsArrayAscending() = fabricLoader.loaders.toTypedArray()

    /**
     * @author Griefed
     */
    actual override fun loaderVersionsArrayDescending() = fabricLoader.loaders.reversed().toTypedArray()

    /**
     * @author Griefed
     */
    actual override fun installerVersionsListAscending() = fabricInstaller.installers

    /**
     * @author Griefed
     */
    actual override fun installerVersionsListDescending() = fabricInstaller.installers.reversed()

    /**
     * @author Griefed
     */
    actual override fun installerVersionsArrayAscending() = fabricInstaller.installers.toTypedArray()

    /**
     * @author Griefed
     */
    actual override fun installerVersionsArrayDescending() = fabricInstaller.installers.reversed().toTypedArray()

    /**
     * @author Griefed
     */
    actual override fun latestInstallerUrl() = fabricInstaller.latestInstallerUrl()

    /**
     * @author Griefed
     */
    actual override fun releaseInstallerUrl() = fabricInstaller.releaseInstallerUrl()

    /**
     * @author Griefed
     */
    actual override fun installerFor(version: String) =
        if (isInstallerUrlAvailable(version)) {
            val destination = File(installerDirectory, "$version.jar")
            if (!destination.isFile) {
                val url = fabricInstaller.installerUrlMeta[version]!!
                val downloaded = utilities.webUtilities.downloadFile(destination, url)
                if (downloaded) {
                    Optional.of(destination)
                } else {
                    Optional.empty()
                }
            } else {
                Optional.of(destination)
            }
        } else {
            Optional.empty()
        }

    /**
     * Download the improved Fabric launcher and store it in the launcher-cache directory.
     * @author Griefed
     */
    actual fun launcherFor(minecraftVersion: String, fabricVersion: String): Optional<File> {
        val destination = File(launchersDirectory, "$minecraftVersion-$fabricVersion.jar")
        return if (!destination.isFile) {
            val url = fabricInstaller.improvedLauncherUrl(minecraftVersion, fabricVersion)
            val downloaded = utilities.webUtilities.downloadFile(destination, url)
            if (downloaded) {
                Optional.of(destination)
            } else {
                Optional.empty()
            }
        } else {
            Optional.of(destination)
        }
    }

    /**
     * @author Griefed
     */
    actual override fun isInstallerUrlAvailable(version: String) =
        Optional.ofNullable(fabricInstaller.installerUrlMeta[version]).isPresent

    /**
     * @author Griefed
     */
    actual override fun getInstallerUrl(version: String) =
        Optional.ofNullable(fabricInstaller.installerUrlMeta[version])

    /**
     * @author Griefed
     */
    actual override fun isVersionValid(version: String) = fabricLoader.loaders.contains(version)

    /**
     * @author Griefed
     */
    actual override fun isMinecraftSupported(minecraftVersion: String) =
        fabricIntermediaries.getIntermediary(minecraftVersion).isPresent

    /**
     * Get the [URL] to the Fabric launcher for the specified Minecraft and Fabric version.
     *
     * @param minecraftVersion Minecraft version.
     * @param fabricVersion    Fabric version.
     * @return URL to the Fabric launcher for the specified Minecraft and Fabric version.
     * @author Griefed
     */
    actual fun improvedLauncherUrl(minecraftVersion: String, fabricVersion: String) =
        fabricInstaller.improvedLauncherUrl(minecraftVersion, fabricVersion)

    /**
     * Get details for a Fabric loader.
     *
     * @param minecraftVersion Minecraft version.
     * @param fabricVersion    Fabric version.
     * @return Details of a Fabric loader for the given Minecraft and Fabric version, wrapped in an
     * [Optional].
     * @author Griefed
     */
    fun getLoaderDetails(
        minecraftVersion: String,
        fabricVersion: String
    ): Optional<FabricDetails> {
        val key = "$minecraftVersion-$fabricVersion"
        return if (loaderDetails.containsKey(key)) {
            Optional.of(loaderDetails[key] as FabricDetails)
        } else if (fabricLoaderDetails.getDetails(minecraftVersion, fabricVersion).isPresent) {
            loaderDetails[key] = fabricLoaderDetails.getDetails(minecraftVersion, fabricVersion).get()
            Optional.of(loaderDetails[key] as FabricDetails)
        } else {
            Optional.empty()
        }
    }
}